assets/app/scripts/directives/fromFile.js

```diff
@@ -0,0 +1,316 @@
+
+angular.module("openshiftConsole")
+  .directive("fromFile", function($q,
+                                  $uibModal,
+                                  $location,
+                                  $filter,
+                                  TemplateService,
+                                  TaskList,
+                                  DataService,
+                                  APIService) {
+    return {
+      restrict: "E",
+      scope: false,
+      templateUrl: "views/directives/from-file.html",
+      controller: function($scope) {
+        var aceEditorSession;
+        var humanize = $filter('humanize');
+        TaskList.clear();
+
+        $scope.aceLoaded = function(editor) {
+          aceEditorSession = editor.getSession();
+          aceEditorSession.setOption('tabSize', 2);
+          aceEditorSession.setOption('useSoftTabs', true);
+          editor.setDragDelay = 0;
+          editor.$blockScrolling = Infinity;
+
+          $('.from-file .editor').animate({
+            height: Math.floor(window.innerHeight * 0.50) + 'px'
+          }, 30);
+        };
+
+        // Determine whats the input format (JSON/YAML) and set appropriate view mode
+        var aceDebounce = _.debounce(function(){
+          try {
+            JSON.parse($scope.editorContent);
+            aceEditorSession.setMode("ace/mode/json");
+          } catch (e) {
+            try {
+              YAML.parse($scope.editorContent);
+              aceEditorSession.setMode("ace/mode/yaml");
+            } catch (e) {
+              return;
+            }
+          }
+        }, 500);
+
+        // Check if the editor isn't empty to disable the 'Add' button. Also check in what format the input is in (JSON/YAML) and change
+        // the editor accordingly.
+        $scope.aceChanged = function() {
+          aceDebounce();
+        };
+
+        $scope.create = function() {
+          delete $scope.alerts['create'];
+          delete $scope.alerts['parsing'];
+          var resource;
+
+          // Trying to auto-detect what format the input is in. Since parsing JSON throws only SyntexError
+          // exception if the string to parse is not valid JSON, it is tried first and then the YAML parser
+          // is trying to parse the string. If that fails it will print the reason. In case the real reason
+          // is JSON related the printed reason will be "Reason: Unable to parse", in case of YAML related
+          // reason the true reason will be printed, since YAML parser throws an error object with needed
+          // data.
+          try {
+            resource = JSON.parse($scope.editorContent);
+          } catch (e) {
+            try {
+              resource = YAML.parse($scope.editorContent);
+            } catch (e) {
+              $scope.error = e;
+              return;
+            }
+          }
+
+          if (!resource.metadata || !resource.kind) {
+            missingFieldsError(resource);
+            return;
+          }
```

Comment:
Maybe have a `validateFields` function that returns an error so we're not checking if these fields exist in two places (here and in the `missingFieldsError` function).